### PR TITLE
Fix race in mark_graph_task_completed.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36640 Fix race in mark_graph_task_completed.**

We had the following race when two threads entered
'mark_graph_task_completed'.

1) Thread 1 grabs the graph_task mutex first and moves captured_vars_ to its
local 'vars'.
2) Thread 1 releases the lock.
3) Thread 2 grabs the mutex and moves an empty captured_vars_ to its local
'vars'.
4) Thread 2 now proceeds to call 'markCompleted' with empty grads.
5) Thread 1 which actually has the right grads never gets to set the grads on
the future since future_completed_ is set to True by Thread 2.

Discovered this while running our RNN example:
https://github.com/pytorch/examples/tree/master/distributed/rpc/rnn and
verified this PR fixes the race.

Differential Revision: [D21035196](https://our.internmc.facebook.com/intern/diff/D21035196/)

### CODE REVIEW

**Review by jjlilley — APPROVED** [review]
Thanks!

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -934,24 +934,27 @@ std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
 }
 
 void Engine::mark_graph_task_completed(const std::shared_ptr<GraphTask>& graph_task) {
-  if (graph_task->future_completed_.load()) {
+  // Allow only one thread one attempt to process this logic.
+  if (graph_task->future_completed_.exchange(true) ||
+      graph_task->future_result_->completed()) {
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:939]: How can the future_result_ be completed without the future_completed_ flag to be already set?

**Review by albanD — COMMENTED** [review]
Nice catch ! Thanks.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (jjlilley) [reply torch/csrc/autograd/engine.cpp:939]: btw, I also expect that this second line isn't necessary.

Comment (pritamdamania87) [reply torch/csrc/autograd/engine.cpp:939]: Good point, I'll remove the extra check.

**Pull request closed without merging**